cassandra/components.py

```diff
@@ -628,16 +645,146 @@ def run_component(self):
+        self.add_args_from_capability(args, 'PrecipitationFile', 'gridded_pr', 'mm_month-1')
+        self.add_args_from_capability(args, 'trn_tas', 'gridded_tas', 'C')
 
         xth_results = xth.execute(args)
 
-        # Add runoff (Q) results from xanthos to the xanthos capability
-        self.addresults("xanthos", xth_results.Q)
+        # Add runoff (Q) results from xanthos to the gridded_runoff capability
+        self.addresults("gridded_runoff", xth_results.Q)
+
+        return 0
+
+    def add_args_from_capability(self, args, arg_name, cap_name, units=None):
+        """Get Xanthos parameters from another capability.
+
+        If the no component provides the requested capability, no arguments are added.
+
+        params:
+          args: Reference to dictionary passed to Xanthos
+          arg_name: Name of argument to add
+          cap_name: Name of capability with required value
+          units: Optional string to use to assert correct units
+        """
+        if cap_name in self.cap_tbl:
+            arg_val = self.fetch(cap_name)
+            assert units is None or arg_val['units'] == units
+            args[arg_name] = self.prep_for_xanthos(arg_val)
+
+    def prep_for_xanthos(self, monthly_data):
+        """Convert a fldgen output array to Xanthos' expected input format.
+
+        Retrieve Xanthos grid cells from alternately indexed vectors.
+
+        Convert spatial data from a list of numeric vectors (each list element is one
+        month) to the input format expected by Xanthos (land cell x months).
+
+        params:
+          monthly_data: Input data for xanthos as numpy array (cells x months)
+
+        returns:
+          2d array of Xanthos cells by month
+
+        """
+        ycolnames = c('index', 'lat', 'lon')
+        ycells = pd.DataFrame(monthly_data['data'].T, columns=ycolnames)
+
+        bothcells = ycells.merge(self.cell_map, on=['lat', 'lon'])
+
+        # Order by Xanthos cell id for indexing
+        bothcells.sort_values(by='cell_id', inplace=True)
+
+        # Re-order the data to correct order by extracting with the cell index
+        cells = monthly_data[:, bothcells.index.astype(int)]
+
+        # Rows and columns need to be flipped
+        return cells.T
+
+
+class FldgenComponent(ComponentBase):
+    """Class for the precipitation and temperature grids
+
+    This component makes use of the fldgen code. That code lives in its
+    own repository and must be installed independently.
+
+    To start, only works with the streamlined branch modified not to
+    call Xanthos.
+
+    params:
+       workdir  - working directory (location of frontEnd_grandExp code)
+
+    results: pr and temp grids
+    """
+
+    def __init__(self, cap_tbl):
+        super(FldgenComponent, self).__init__(cap_tbl)
+        self.addcapability("gridded_pr")
+        self.addcapability("gridded_tas")
+
+    def run_component(self):
+        """Run the fldgen R scripts."""
```

Comment:
It looks like this is running both fldgen and an2month, which is fine, but we should be a little more clear about that fact so that when we're replacing an2month somewhere down the line we don't have a bunch of confusion.

```suggestion
        """Run the fldgen and an2month R scripts."""
```